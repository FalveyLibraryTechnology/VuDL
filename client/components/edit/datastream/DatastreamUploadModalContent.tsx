import React from "react";
import { Box, DialogContent } from "@mui/material";
import { useState } from "react";
import { useEditorContext } from "../../../context/EditorContext";
import { useFetchContext } from "../../../context/FetchContext";
import { postObjectDatastreamUrl } from "../../../util/routes";

const DatastreamUploadModalContent = () => {
    const {
        action: { fetchText },
    } = useFetchContext();
    const {
        state: { currentPid, activeDatastream, datastreamsCatalog },
    } = useEditorContext();
    const [{ state, text }, setResponseText] = useState({
        state: "init",
        text: "",
    });
    const isAllowedMimeType = (mimeType) => {
        if (!datastreamsCatalog[activeDatastream]) {
            return false;
        }
        const [type, subtype] = mimeType.split("/");
        const { allowedType, allowedSubtypes } = datastreamsCatalog[activeDatastream].mimetype;
        return (
            (allowedType.includes(type) || allowedType.includes("*")) &&
            (allowedSubtypes.includes(subtype) || allowedSubtypes.includes("*"))
        );
    };

    const uploadFile = async (file) => {
        try {
            if (!isAllowedMimeType(file.type)) {
                throw new Error(`Illegal mime type: ${file.type}`);
            }
            const body = new FormData();
            body.append("stream", activeDatastream);
            body.append("file", file);
            const text = await fetchText(postObjectDatastreamUrl(currentPid), {
                method: "POST",
                body,
            });
            setResponseText({
                state: "success",
                text,
            });
        } catch (err) {
            setResponseText({
                state: "error",
                text: err.message,
            });
        }
    };
    const onChange = (event) => {
        uploadFile(event.target.files[0]);
    };

    return (
        <DialogContent>
            <input className="uploadFileButton" type="file" onChange={onChange} />
            {state != "init" && (
                <Box className="responseText" sx={{ color: `${state}.main`, textAlign: "center" }}>
                    {text}
                </Box>
            )}
        </DialogContent>
    );
};

export default DatastreamUploadModalContent;
