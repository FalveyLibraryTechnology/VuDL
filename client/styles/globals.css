:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
