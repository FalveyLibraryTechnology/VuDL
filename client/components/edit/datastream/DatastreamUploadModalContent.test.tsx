import React from "react";
import { describe, beforeEach, expect, it, jest } from "@jest/globals";
import { mount, shallow } from "enzyme";
import { act } from "react-dom/test-utils";
import toJson from "enzyme-to-json";
import DatastreamUploadModalContent from "./DatastreamUploadModalContent";

const mockUseEditorContext = jest.fn();
jest.mock("../../../context/EditorContext", () => ({
    useEditorContext: () => {
        return mockUseEditorContext();
    },
}));
const mockUseFetchContext = jest.fn();
jest.mock("../../../context/FetchContext", () => ({
    useFetchContext: () => {
        return mockUseFetchContext();
    },
}));
describe("DatastreamUploadModalContent", () => {
    let fetchValues;
    let editorValues;
    beforeEach(() => {
        fetchValues = {
            action: {
                fetchText: jest.fn(),
            },
        };
        editorValues = {
            action: {
                getCurrentModelsDatastreams: jest.fn().mockResolvedValue({}),
            },
            state: {
                currentPid: "vudl:123",
                activeDatastream: "THUMBNAIL",
                datastreamsCatalog: {
                    THUMBNAIL: {
                        mimetype: {
                            allowedType: "image",
                            allowedSubtypes: "png",
                        },
                    },
                },
            },
        };
        mockUseEditorContext.mockReturnValue(editorValues);
        mockUseFetchContext.mockReturnValue(fetchValues);
    });

    it("renders", () => {
        const wrapper = shallow(<DatastreamUploadModalContent />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders upload success", async () => {
        fetchValues.action.fetchText.mockResolvedValue("upload worked");
        const wrapper = mount(<DatastreamUploadModalContent />);
        await act(async () => {
            wrapper.find(".uploadFileButton").simulate("change", {
                target: {
                    files: [
                        {
                            type: "image/png",
                        },
                    ],
                },
            });
            wrapper.update();
        });

        expect(fetchValues.action.fetchText).toHaveBeenCalledWith(
            expect.stringContaining(editorValues.state.currentPid),
            expect.objectContaining({
                method: "POST",
                body: expect.any(FormData),
            })
        );
        expect(editorValues.action.getCurrentModelsDatastreams).toHaveBeenCalled();
        expect(wrapper.children().text()).toEqual("upload worked");
    });

    it("renders illegal mime type when type is invalid", async () => {
        const wrapper = mount(<DatastreamUploadModalContent />);
        await act(async () => {
            wrapper.find(".uploadFileButton").simulate("change", {
                target: {
                    files: [
                        {
                            type: "image/illegaltype",
                        },
                    ],
                },
            });
            wrapper.update();
        });

        expect(wrapper.children().text()).toContain("Illegal mime type");
    });

    it("renders illegal mime type when catalog cannot find datastream", async () => {
        editorValues.state.datastreamsCatalog = {};
        const wrapper = mount(<DatastreamUploadModalContent />);
        await act(async () => {
            wrapper.find(".uploadFileButton").simulate("change", {
                target: {
                    files: [
                        {
                            type: "image/png",
                        },
                    ],
                },
            });
            wrapper.update();
        });

        expect(wrapper.children().text()).toContain("Illegal mime type");
    });
});
