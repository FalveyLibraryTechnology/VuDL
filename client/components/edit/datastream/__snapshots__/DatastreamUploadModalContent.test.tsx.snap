// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatastreamUploadModalContent renders 1`] = `
<ForwardRef(DialogContent)>
  <input
    className="uploadFileButton"
    onChange={[Function]}
    type="file"
  />
</ForwardRef(DialogContent)>
`;
